Objects/regions.c

```diff
@@ -94,9 +120,299 @@ struct regionmetadata {
+    // increased the rc.
+    regionmetadata_dec_rc(Py_region_ptr(self->parent));
+
+    free(self);
+}
+#define regionmetadata_dec_rc(self) \
+    (regionmetadata_dec_rc(REGION_PTR_CAST(self)))
+
+static void regionmetadata_set_parent(regionmetadata* self, regionmetadata* parent) {
+    // Just a sanity check, since these cases should never happen
+    assert(HAS_METADATA(self) && "Can't set the parent on the immutable and local region");
+    assert(REGION_PTR_CAST(self) == regionmetadata_get_root(self) && "Sanity Check");
+    assert(REGION_PTR_CAST(parent) == regionmetadata_get_root(parent) && "Sanity Check");
+
+    Py_region_ptr_t old_parent = Py_region_ptr(self->parent);
+    Py_region_ptr_t new_parent = REGION_PTR_CAST(parent);
+    self->parent = Py_region_ptr_with_tags(new_parent);
+
+    // Update RCs
+    regionmetadata_inc_rc(new_parent);
+    if (regionmetadata_is_open(self)) {
+        regionmetadata_inc_osc(new_parent);
+        regionmetadata_dec_osc(old_parent);
+    }
+    regionmetadata_dec_rc(old_parent);
+}
+
+static regionmetadata* regionmetadata_get_parent(regionmetadata* self) {
+    assert(REGION_PTR_CAST(self) == regionmetadata_get_root(self) && "Sanity check");
+    if (!HAS_METADATA(self)) {
+        // The local and immutable regions never have a parent
+        return NULL;
+    }
+
+    Py_region_ptr_t parent_field = Py_region_ptr(self->parent);
+    Py_region_ptr_t parent_root = regionmetadata_merge_root(parent_field);
+
+    // If the parent was merged with another region we want to update the
+    // pointer to point at the root.
+    if (parent_field != parent_root) {
+        // set_parent ensures that the RC's are correctly updated
+        regionmetadata_set_parent(self, REGION_DATA_CAST(parent_root));
+    }
+
+    return REGION_DATA_CAST(parent_root);
+}
+#define regionmetadata_get_parent(self) \
+    regionmetadata_get_parent(REGION_DATA_CAST(self))
+
+static bool regionmetadata_has_parent(regionmetadata* self) {
+    return regionmetadata_get_parent(self) != NULL;
+}
+
+static bool regionmetadata_has_ancestor(regionmetadata* self, regionmetadata* other) {
+    // The immutable or local region can never be a parent
+    if (!HAS_METADATA(other)) {
+        return false;
+    }
+
+    while (self) {
+        if (self == other) {
+            return true;
+        }
+        self = regionmetadata_get_parent(self);
+    }
+    return false;
+}
+
+
+// This implementation merges `self` into `other`. Merging is not allowed
+// to break external uniqueness. It's therefore not allowed if both regions
+// to have a parent. Except cases, where one region has the other region as
+// it's parent.
+//
+// This function expects `self` to be a valid object.
+__attribute__((unused))
+static PyObject* regionmetadata_merge(regionmetadata* self, Py_region_ptr_t other) {
+    assert(HAS_METADATA(self) && "The immutable and local region can't be merged into another region");
```

Comment:
The underlying assumptions of this error message may not be true -- self may be a cown etc. Also, if self and other are both immutable the message is misleading.

Reply:
> self may be a cown etc.

I assume you mean the cown region and not an actual cown object?

Even then, there is the question when this can happen? The type of `self` is `regionmetadata*` which disallowed `Py_IMMUTABLE` and `Py_COWN` unless someone casts incorrectly in which case the assert would complain.

There should never be a case where we try to merge `Py_IMMUTABLE`, `Py_LOCAL` or `Py_COWN` **into** another region. While other regions can be **merged into** those.

This restriction can be easily lifted. I just think that it makes it easier to have this as a "*merge self into other*" function which catches accidental tries of merging `Py_IMMUTABLE`, `Py_LOCAL` or `Py_COWN` into an existing region.